### DISCUSSION TITLE
How to wrap a static constexpr data member (an integer value), for Python?

### DISCUSSION THREAD

**Discussion by N-Dekker** [swig/swig#3088] [General]
Suppose we have a class template:

```cpp

template <int N> class Point
{
public:
  static constexpr int Dimension = N;
  // ...
};
```
(A simplified version of [itk::Point](https://github.com/InsightSoftwareConsortium/ITK/blob/4cbe24cb4a45d689cadd56d554b8ccf3584a5ca6/Modules/Core/Common/include/itkPoint.h))

Is it possible to retrieve the value of the static constexpr data member `Dimension` in Python, from an instance of `Point`, wrapped by SWIG? For example as in:

```py
print(point.Dimension)
print(type(point).Dimension)
```

In Python, I get errors like:
> AttributeError: ... has no attribute 'Dimension'

**Comment by wsfulton** [comment]
Please read the docs about templates. Looks like you are missing an instantiation using `%template`.

**Comment by N-Dekker** [comment]
Thanks @wsfulton ! I finally had time to figure out how to manually wrap such a simple `Point<N>` template class, using Visual C++. And it works, indeed 👍 I did so by adding the template to the "example.h" in [Examples\python\template](https://github.com/swig/swig/tree/8fd4df01bd4111e5f7322fe4f9d4817ffd6f0c6d/Examples/python/template), and by adding those two `%template` instantiations to "example.i": 

```
%template(Point2) Point<2>;
%template(Point3) Point<3>;
```

Then eventually in Python I could do:

```python
>>> print(example.Point2.Dimension)
2
>>> print(example.Point3.Dimension)
3
```

Unfortunately I still don't know why it does not work for the Python wrapping of ITK (for example, [itk::Point](https://github.com/InsightSoftwareConsortium/ITK/blob/4cbe24cb4a45d689cadd56d554b8ccf3584a5ca6/Modules/Core/Common/include/itkPoint.h)). As observed by Matt McCormick (@thewtex) at https://github.com/InsightSoftwareConsortium/ITK/pull/5045/#discussion_r1882153159 So it still isn't clear to me 🤷

**Comment by N-Dekker** [comment]
ITK's template (struct) `itk::Size<unsigned int VDimension>` appears to have the same problem. It is implemented at [ITK/Modules/Core/Common/include/itkSize.h](https://github.com/InsightSoftwareConsortium/ITK/blob/2fd36ab3a6871340c1defd5a69dc657a81bb8134/Modules/Core/Common/include/itkSize.h
) 

[itkSize.h line 83](https://github.com/InsightSoftwareConsortium/ITK/blob/2fd36ab3a6871340c1defd5a69dc657a81bb8134/Modules/Core/Common/include/itkSize.h#L83) defines the public static data member `Dimension`, as follows: 

```cpp
  static constexpr unsigned int Dimension = VDimension;
```

The SWIG-based wrapping of ITK allows creating an `itk::Size<2>` object (for example, `size_2d`) in Python as follows:

```py
>>> import itk
>>> size_2d = itk.Size[2]()
>>> print(type(size_2d))
<class 'itk.itkSizePython.itkSize2'>
>>> 
```
This `size_2d` has the methods wrapped from the `itk::Size<2>` implementation (`CalculateProductOfElements`, `Fill`, `Filled`, etc), but apparently, `itk::Size<2>::Dimension` is not wrapped:
```py
>>> print(dir(size_2d))
['CalculateProductOfElements', 'Fill', 'Filled', 'GetElement', 'GetSize', 'GetSizeDimension', 'SetElement', 'SetSize', '__add__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__isub__', '__le__', '__len__', '__lt__', '__module__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__swig_destroy__', '__weakref__', 'assign', 'at', 'back', 'begin', 'cbegin', 'cend', 'data', 'empty', 'end', 'front', 'max_size', 'rbegin', 'rend', 'size', 'swap', 'this', 'thisown']
>>> print(size_2d.Dimension) 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'itkSize2' object has no attribute 'Dimension'
```

Do you have any clue?

**Reply by N-Dekker** [reply]
@blowekamp Hi Bradley! As you're an expert in "swigging" ITK, being Mr. [SimpleITK](https://github.com/SimpleITK), can you possibly also have a look here as well? It appears that SWIG is very well able to wrap public static constexpr data members, but not for ITK! As far as I can see 🤷 Do you have a clue?

**Reply by blowekamp** [reply]
The first step would be to look at the SWIG interface  file ( e.g. itkSize.i) generated by the ITK build system to determine which directive are given.

**Reply by N-Dekker** [reply]
Thanks @blowekamp I found it in my binary (build) directory, `<Bin>\ITK\Wrapping\Typedefs\itkSize.i`.  It contains class definitions of itkSize2, itkSize3, and itkSize4, but none of them still has a static data member named `Dimension`! (Even though it was defined in the original source file at [itkSize.h line 83](https://github.com/InsightSoftwareConsortium/ITK/blob/2fd36ab3a6871340c1defd5a69dc657a81bb8134/Modules/Core/Common/include/itkSize.h#L83).) How can that be?

For example:

```cpp
class itkSize2 {
  public:
    static unsigned int GetSizeDimension();
    itkSize2 const operator+(itkSize2 const & vec) const;
    itkSize2 const & operator+=(itkSize2 const & vec);
    itkSize2 const operator-(itkSize2 const & vec) const;
    itkSize2 const & operator-=(itkSize2 const & vec);
    itkSize2 const operator*(itkSize2 const & vec) const;
    itkSize2 const & operator*=(itkSize2 const & vec);
    long unsigned long const * GetSize() const;
    void SetSize(long unsigned long const * val);
    void SetElement(unsigned long element, long unsigned long val);
    long unsigned long GetElement(unsigned long element) const;
    void Fill(long unsigned long value);
    long unsigned long CalculateProductOfElements() const;
    void assign(long unsigned long const & newValue);
    void swap(itkSize2 & other);
    long unsigned long const * cbegin() const;
    long unsigned long * begin();
    long unsigned long const * begin() const;
    long unsigned long const * cend() const;
    long unsigned long * end();
    long unsigned long const * end() const;
    std::reverse_iterator< unsigned long long * > rbegin();
    std::reverse_iterator< const unsigned long long * > rbegin() const;
    std::reverse_iterator< unsigned long long * > rend();
    std::reverse_iterator< const unsigned long long * > rend() const;
    unsigned int size() const;
    unsigned int max_size() const;
    bool empty() const;
    long unsigned long & operator[](unsigned int pos);
    long unsigned long const & operator[](unsigned int pos) const;
    long unsigned long & at(unsigned int pos);
    long unsigned long const & at(unsigned int pos) const;
    long unsigned long & front();
    long unsigned long const & front() const;
    long unsigned long & back();
    long unsigned long const & back() const;
    long unsigned long * data();
    long unsigned long const * data() const;
    static itkSize2 Filled(long unsigned long const value);
    itkSize2();
    itkSize2(itkSize2 const & arg0);
    itkSize2 & operator=(itkSize2 const & arg0);
    ~itkSize2();
  private:
    void ExceptionThrowingBoundsCheck(unsigned int pos) const;
  protected:
};
```

**Reply by N-Dekker** [reply]
Thanks @blowekamp I understand now that itkSize.i is generated by ITK's [igenerator.py](https://github.com/InsightSoftwareConsortium/ITK/blob/2fd36ab3a6871340c1defd5a69dc657a81bb8134/Wrapping/Generators/SwigInterface/igenerator.py), which appears to skip static constexpr data members, unfortunately. But anyway, SWIG does properly support static constexpr data members, so it's not a SWIG issue.

**Reply by N-Dekker** [reply]
@blowekamp Instead of having those class definitions for itkSize2, itkSize3, and itkSize4, shouldn't the generated "itkSize.i" file just have SWIG template instantiations of the form:

```cpp
%template(itkSize2) itk::Size<2>;
%template(itkSize3) itk::Size<3>;
%template(itkSize4) itk::Size<4>;

```
? 

Just wondering, could it be that SWIG's `%template` wasn't yet available when ITK's igenerator.py was developed? (The original [igenerator](https://github.com/InsightSoftwareConsortium/ITK/blob/0248e8fa1cb8b547cc04f84e708011c696518c23/Wrapping/WrapITK/Languages/SwigInterface/igenerator.py.in) was added to ITK in January 2011.) Or is there a fundamental reason that ITK needs something more than just `%template`?

**Reply by blowekamp** [reply]
This is clearly an ITK issue, please create an issue there were we can continue the conversation, and keep noise away from the SWIG developers.

**Reply by N-Dekker** [reply]
OK, you're right @blowekamp I'll close the discussion here, and follow-up at https://discourse.itk.org/t/why-doesnt-itk-swiginterface-igenerator-py-just-use-swig-template/7377
